staging/src/k8s.io/client-go/transport/cache.go

```diff
@@ -138,6 +147,21 @@ func (c *tlsTransportCache) get(config *Config) (http.RoundTripper, error) {
 	return transport, nil
 }
 
+type upgradeAwareTransport struct {
+	h1, h2 http.RoundTripper
+}
+
+func (tr *upgradeAwareTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if v := req.Header.Get("Upgrade"); v != "" {
+		return tr.h1.RoundTrip(req)
+	}
+	return tr.h2.RoundTrip(req)
+}
+
+func (tr *upgradeAwareTransport) WrappedRoundTripper() http.RoundTripper {
+	return tr.h2
+}
```

Comment:
This does not look correct to me (at least not in all contexts).  Looking through the usages of `WrappedRoundTripper`, we use it in a few places, one of them being request cancellation.  In such cases, the correct round tipper to return is based on the request.

Reply:
Pushed different implementation of this. PTAL.